shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -164,52 +164,114 @@ class JSBackend {
       case NoCases        => Ls(JSThrowStmt())
     }
 
-  private val resolvedAliases = collection.mutable.HashMap[Str, Str]()
+  private val typeAliasMap =
+    collection.mutable.HashMap[Str, Ls[Str] -> Type]()
 
-  private def resolveAliases(typeDefs: Ls[TypeDef]) = {
-    val relations = collection.immutable.HashMap.from(typeDefs flatMap {
-      case TypeDef(Als, Primitive(alias), _, Primitive(name)) =>
-        S(alias -> name)
-      case _ => N
-    })
-    resolvedAliases.addAll(
-      Ls.from(relations flatMap { case (s, t) => s :: t :: Nil }).distinct map {
-        name =>
-          val visited = HashSet.empty[Str]
-          var current = name
-          while (!visited.contains(current)) {
-            visited += current
-            relations.get(current) match {
-              case N =>
-              case S(alias) =>
-                current = alias
+  // This function evaluates something like `T[A, B]`.
+  private def applyTypeAlias(name: Str, targs: Ls[Type]): Type =
+    typeAliasMap get name match {
+      case S(tparams -> body) =>
+        if (targs.length === tparams.length) {
+          val subs = collection.immutable.HashMap(
+            tparams zip targs map { case (k, v) => k -> v }: _*
+          )
+          substitute(body, subs)
+        } else {
```

Comment:
This whole verbose conditional should probably just be an `assert(targs.length === tparams.length, targs -> tparams)` at the beginning.